File: 700/757 - Set Intersection Size At Least Two.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public int intersectionSizeTwo(int[][] intervals) {
        List<Integer> s = new ArrayList<>();
        s.add(-1);
        s.add(-1);
        
        Arrays.sort(intervals, (int[] a, int[] b) -> {
            return (a[1] > b[1] || (a[1] == b[1] && a[0] < b[0])) ? 1 : -1;
        });
        
        for (int[] i : intervals) {
            int len = s.size();
            if (i[0] <= s.get(len - 2)) {
                continue;
            }
            
            if (i[0] > s.get(len - 1)) {
                s.add(i[1] - 1);
            }
            s.add(i[1]);
        }

        return s.size() - 2;
    }
}

File: csharp/757.set-intersection-size-at-least-two.cs
/*
 * @lc app=leetcode id=757 lang=csharp
 *
 * [757] Set Intersection Size At Least Two
 */

using System.Collections.Generic;
using System;

// @lc code=start
public class Solution {
    public int IntersectionSizeTwo(int[][] intervals) {
        int ret = 0, pp = -1, p = -1;
        Array.Sort(intervals, Comparer<int[]>.Create((a, b) => a[1] == b[1] ? b[0].CompareTo(a[0]) : a[1].CompareTo(b[1])));
        foreach (var interval in intervals) {
            if (interval[0] <= pp) {
                continue;
            }
            if (interval[0] > p) {
                ret += 2;
                pp = interval[1] - 1;
                p = interval[1];
            } else {
                ++ret;
                pp = p;
                p = interval[1];
            }
        }
        return ret;
    }
}
// @lc code=end

